plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/storage/TrinoHudiStorage.java

```diff
@@ -195,16 +195,28 @@ public boolean createDirectory(StoragePath path)
     public List<StoragePathInfo> listDirectEntries(StoragePath path)
             throws IOException
     {
-        FileIterator fileIterator = fileSystem.listFiles(convertToLocation(path));
+        Location location = convertToLocation(path);
+        Optional<Boolean> dirExists = fileSystem.directoryExists(location);
+
+        // For hierarchical filesystems (HDFS): directoryExists returns true/false
+        // For non-hierarchical filesystems (S3): returns empty if directory doesn't exist
```

Comment:
Does it means we can throw here as well if dirExists is empty?

Reply:
No, we shouldn't throw immediately when dirExists is empty. On S3, directoryExists() returns Optional.empty() when it can't determine existence without scanning. A directory like s3://bucket/a/b/ can implicitly exist through files like s3://bucket/a/b/c.txt
   even without an explicit directory marker.

  If we throw early, we'd incorrectly reject valid S3 directories that contain files but lack explicit markers. We need to attempt listFiles() first and only throw if both the list is empty AND dirExists.isEmpty() (lines 216-218).